### PR TITLE
Fix for plot script unhiding other plots (switch plt.show() to fig.show())

### LINKED ISSUE #34978: Executing a plot script will show all hidden plots
**Describe the bug**
When executing a script that plots some figure, all hidden plots will be shown with the figure executed from the script. Also, if the figure is already present in the hidden menu, it will be plotted with an extra legend. 

**To Reproduce**
1- Load a workspace
2- Plot the data
3- Hide the plot
4- Load another workspace
5- Plot the second workspace
6- Copy the python script of the second plot to the clipboard
7- Hide the second plot
8- Paste the script into the editor
9- Execute the script
10- You will see two plots, the first workspace plot with two legends and the second workspace plot

**Expected behavior**

When executing the script it should only show one plot with one legend.

**Screenshots**

**Platform/Version (please complete the following information):**
 - OS: Windows
 - OS Version: 10 Enterprise
 - Mantid Version: 6.6.20230115.2237

**Additional context**
The problem can be fixed if plt is changed to fig at the end of the script.

### LINKED ISSUE #36938: Mantid Figure/plot script generator uses plt.show() not fig.show()
Original reporter: Pascal Manuel

**Describe the bug**
Mantid Figure/plot script generator uses `plt.show()` not `fig.show()`
It's been reported by users (incl. Pascal on WISH) that `plt.show()` causes mantid to hang.

**To Reproduce**
(1) `ws = CreateSampleWorkspace()`
(2) Plot a spectrum of `ws`
(3) Copy script to clipboard using button in toolbar

Produces 
```
fig, axes = plt.subplots(edgecolor='#ffffff', num='ws-3', subplot_kw={'projection': 'mantid'})
axes.plot(ws, color='#1f77b4', label='ws: spec 1', wkspIndex=0)
axes.tick_params(axis='x', which='major', **{'gridOn': False, 'tick1On': True, 'tick2On': False, 'label1On': True, 'label2On': False, 'size': 6, 'tickdir': 'out', 'width': 1})
axes.tick_params(axis='y', which='major', **{'gridOn': False, 'tick1On': True, 'tick2On': False, 'label1On': True, 'label2On': False, 'size': 6, 'tickdir': 'out', 'width': 1})
axes.set_title('ws')
axes.set_xlabel('Time-of-flight ($\\mu s$)')
axes.set_ylabel('Counts ($\\mu s$)$^{-1}$')
legend = axes.legend(fontsize=8.0).set_draggable(True).legend

plt.show()
```
not 
```
fig.show()
```

**Expected behavior**
Would have last line `fig.show()`

### PR DESCRIPTION
**Pull request by jclarkeSTFC**
### Description of work
The script generated by a plot window ends with a call to `plt.show()`, but `show()` causes all hidden plots to be unhidden. This was changed from calling `fig.show()` in #28699 because the `matplotlib` online documentation usually uses `plt.show()`. There's an explanation [here](https://stackoverflow.com/a/51570439) about the difference (assuming the random person who wrote that answer is correct). If someone is running the plot script outside of Workbench then `plt.show()` is more appropriate, but within Workbench `fig.show()` is better. I'm not sure which is the most likely scenario, maybe we should leave the code as `plt.show()`. If you call `plt.show()` when running a script outside of Workbench, then the figure window will appear, but if you call `fig.show()` it won't.

Fixes #34978

This has been reopened to respond to #36938. I've added a commented out `plt.show` as suggested. - Jonathan

Fixes #36938

### To test:
See #34978 for how to reproduce. Maybe trying a few different plot types and their generated scripts would be good. I tried a few different ones and they seemed okay.

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Comment by adriazalvarez** [comment]
I think is the best solution to the problem described in #34978. But as noted by James, fixing it in this way on the workbench backend will make things worse for people using the script externally, as fig.show() does not execute its own event loop. 
https://github.com/matplotlib/matplotlib/blob/325fcbbc3e626052ceb8672e80b51abd6a05f94b/lib/matplotlib/figure.py#L2641

I have tried to open an script using this solution on a jupyter notebook, and it works only because the notebook executes plt.show() by default on each cell. (on pycharm for example it does not work). 

We could add a commented 'plt.show()'  in the line next to fig.show() with a  warning to the users of the issue. Otherwise is a use case scenario, if most people will execute the script withing the workbench, then the changes are fine. 

Since this is a draft, I guess is better if somebody with more authority than me comment on it.

**Comment by jclarkeSTFC** [comment]
Probably best to leave it as it is for now, can always come back to this later.

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py [hunk 1]
```diff
@@ -65,7 +65,9 @@ def generate_script(fig, exclude_headers=False):
         axes.legend().set_draggable(True)     (if legend present, or just draggable() for earlier matplotlib versions)
         <Legend title and label commands if non-default>
 
-        plt.show()
+        fig.show()
+        # User plt.show() if running the script outside of Workbench
```
Comment (jclarkeSTFC) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py:69]:
```suggestion
        # Use plt.show() if running the script outside of Workbench
```

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py [hunk 2]
```diff
@@ -42,10 +42,10 @@
     "axes.set_ylim(...)\n"
     "axes.set_facecolor('#8a9aff')\n"
     "\n"
-    "plt.show()"
-    "\n"
-    "# Scripting Plots in Mantid:"
-    "\n"
+    "fig.show()\n"
+    "# User plt.show() if running the script outside of Workbench\n"
```
Comment (jclarkeSTFC) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py:46]:
```suggestion
    "# Use plt.show() if running the script outside of Workbench\n"
```

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py (hunk 1 above)
Comment (warunawickramasingha) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py]: I guess there is a typo in the comment where `User` need to be `Use`. Unfortunately that change requires update to unit tests as well.

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py [hunk 3]
```diff
@@ -71,10 +71,10 @@
...
```
Comment (jclarkeSTFC) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py:75]:
```suggestion
    "# Use plt.show() if running the script outside of Workbench\n"
```

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py [hunk 4]
```diff
@@ -98,7 +98,9 @@
     "axes[1].set_ylim(...)\n"
     "axes[1].set_facecolor('#8a9aff')\n"
     "\n"
-    "plt.show()\n"
+    "fig.show()\n"
+    "# User plt.show() if running the script outside of Workbench\n"
```
Comment (jclarkeSTFC) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/test/test_plotscriptgenerator.py:102]:
```suggestion
    "# Use plt.show() if running the script outside of Workbench\n"
```

File: qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py [hunk 5]
```diff
@@ -122,7 +124,9 @@ def generate_script(fig, exclude_headers=False):
     cmds.extend(generate_workspace_retrieval_commands(fig) + [""])
     cmds.append("{}, {} = {}".format(FIG_VARIABLE, AXES_VARIABLE, generate_subplots_command(fig)))
     cmds.extend(plot_commands)
-    cmds.append("plt.show()")
+    cmds.append("fig.show()")
+    cmds.append("# User plt.show() if running the script outside of Workbench")
```
Comment (warunawickramasingha) [inline qt/applications/workbench/workbench/plotting/plotscriptgenerator/__init__.py:128]:
```suggestion
    cmds.append("# Use plt.show() if running the script outside of Workbench")
```

**Review by warunawickramasingha — CHANGES_REQUESTED** [review]

**Review by warunawickramasingha — APPROVED** [review]
Happy to approve and the fix is working fine.

**Pull request merged**